src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,130 @@
+using Microsoft.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Globalization;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal sealed class W3CLogger : IDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+        private readonly W3CLoggingFields _loggingFields;
+
+        internal W3CLogger(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            _options = options;
+
+            _messageQueue = new W3CLoggerProcessor(_options);
+            _loggingFields = _options.CurrentValue.LoggingFields;
+            LogFileFullName = _messageQueue.FullName;
+        }
+
+        public string? LogFileFullName { get; }
+
+        public void Dispose() => _messageQueue?.Dispose();
+
+        public void Log(IReadOnlyCollection<KeyValuePair<string, object?>> state)
+        {
+            _messageQueue.EnqueueMessage(Format(state));
+        }
+
+        private string Format(IEnumerable<KeyValuePair<string, object?>> state)
+        {
+            // Subtract 1 to account for the "All" flag
+            string[] elements = new string[Enum.GetValues(typeof(W3CLoggingFields)).Length - 1];
+            foreach(KeyValuePair<string, object?> kvp in state)
+            {
+                var val = kvp.Value is null ? "" : kvp.Value.ToString();
+                if (val is null)
+                {
+                    val = "";
+                }
+                switch (kvp.Key)
+                {
+                    case nameof(HttpRequest.Method):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Method)] = val.Trim();
+                        break;
+                    case nameof(HttpRequest.Path):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UriStem)] = val.Trim();
+                        break;
+                    case nameof(HttpRequest.QueryString):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UriQuery)] = val.Trim();
+                        break;
+                    case nameof(HttpResponse.StatusCode):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ProtocolStatus)] = val.Trim();
+                        break;
+                    case nameof(HttpRequest.Protocol):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ProtocolVersion)] = val.Trim();
+                        break;
+                    case nameof(HeaderNames.Host):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Host)] = val.Trim();
+                        break;
+                    case "User-Agent":
+                        // User-Agent can have whitespace - we replace whitespace characters with the '+' character
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UserAgent)] = Regex.Replace(val.Trim(), @"\s", "+");
```

Comment:
Regex 😨

Reply:
What would be better?

Reply:
A custom method with a for loop. It checks each char to see if its space, tab, CR, LF and replaces matches with +.

It is such a common thing that I imagine there is a prior implementation floating around somewhere.

Reply:
As to why regexes are bad, they're most likely slower than doing it yourself.

Also, there are security worries about regexes going rogue which is why regexes have a timeout argument that you're generally supposed to use. This one never could, but something to consider with regexes.

Reply:
Yes, no regexs 😄